pkg/pillar/cmd/zedagent/attesttask.go

```diff
@@ -24,19 +32,197 @@ type WatchdogImpl struct{}
+		}
+		pcrValue.Value = pcr.Digest
+		quoteMsg.PcrValues = append(quoteMsg.PcrValues, pcrValue)
+	}
+	//XXX Check for TPM platform, and if so, insist on non-empty quoteMsg.PCRValues
 	return nil
 }
 
 //SendAttestQuote implements SendAttestQuote method of zattest.Verifier
 func (server *VerifierImpl) SendAttestQuote(ctx *zattest.Context) error {
-	//XXX: Fill it in when Controller code is ready
+	if ctx.OpaqueCtx == nil {
+		log.Fatalf("[ATTEST] Uninitialized access to OpaqueCtx")
+	}
+	attestCtx, ok := ctx.OpaqueCtx.(*attestContext)
+	if !ok {
+		log.Fatalf("[ATTEST] Unexpected type from opaque ctx: %T",
+			ctx.OpaqueCtx)
+	}
+	var attestReq = &attest.ZAttestReq{}
+
+	// bail if V2API is not supported
+	if !zedcloud.UseV2API() {
+		return zattest.ErrControllerReqFailed
+	}
+
+	attestReq.ReqType = attest.ZAttestReqType_ATTEST_REQ_QUOTE
+	//XXX Fill GPS info, Version, Eventlog fields later
+	quote := &attest.ZAttestQuote{
+		AttestData: attestCtx.InternalQuote.Quote,
+		Signature:  attestCtx.InternalQuote.Signature,
+	}
+
+	if err := getPCRValues(attestCtx.InternalQuote, quote); err != nil {
+		log.Errorf("[ATTEST] getPCRValues failed with err %v", err)
+		return zattest.ErrControllerReqFailed
+	}
+
+	attestReq.Quote = quote
+	log.Debugf("Sending Quote request %v", attestReq)
+	_, contents, senderStatus, err := trySendToController(attestReq, attestCtx.Iteration)
+	if err != nil || senderStatus != types.SenderStatusNone {
+		log.Errorf("[ATTEST] Error %v, senderStatus %v",
+			err, senderStatus)
+		//Increment Iteration for interface rotation next time
+		attestCtx.Iteration++
+		return zattest.ErrControllerReqFailed
+	}
+
+	//Reset Iteration on successful reply, irrespective of the reply content
+	attestCtx.Iteration = 0
+
+	attestResp := &attest.ZAttestResponse{}
+	if err := proto.Unmarshal(contents, attestResp); err != nil {
+		log.Errorf("[ATTEST] Error %v in Unmarshaling nonce response", err)
+		return zattest.ErrControllerReqFailed
+	}
+
+	respType := attestResp.GetRespType()
+	if respType != attest.ZAttestRespType_ATTEST_RESP_QUOTE_RESP {
+		log.Errorf("[ATTEST] Got %v, but want %v",
+			respType, attest.ZAttestRespType_ATTEST_RESP_QUOTE_RESP)
+		return zattest.ErrControllerReqFailed
+	}
+
+	var quoteResp *attest.ZAttestQuoteResp
+	if quoteResp = attestResp.GetQuoteResp(); quoteResp == nil {
+		log.Errorf("[ATTEST] Got empty quote response")
+		return zattest.ErrControllerReqFailed
+	}
+	quoteRespCode := quoteResp.GetResponse()
+	switch quoteRespCode {
+	case attest.ZAttestResponseCode_Z_ATTEST_RESPONSE_CODE_INVALID:
+		log.Errorf("[ATTEST] Invalid response code")
+		return zattest.ErrControllerReqFailed
+	case attest.ZAttestResponseCode_Z_ATTEST_RESPONSE_CODE_SUCCESS:
+		//XXX Retrieve Storage keys, and integrity token
+		log.Infof("[ATTEST] Attestation successful")
+		return nil
+	case attest.ZAttestResponseCode_Z_ATTEST_RESPONSE_CODE_NONCE_MISMATCH:
+		log.Errorf("[ATTEST] Nonce Mismatch")
+		return zattest.ErrNonceMismatch
```

Comment:
When we get a nonce mismatch error, does the fsm send a new nonce request? Need for rate limiting in doing that?

Reply:
When there is a nonce mismatch, fsm enters into a wait for 15 seconds before trying again (through EventRestart)